src/ispc/mod.rs

```diff
@@ -1,4 +1,62 @@
 #![allow(deref_nullptr)]
 
 use ispc_rt::ispc_module;
+use std::rc::Rc;
+
+use crate::CachedWeight;
 ispc_module!(downsample_ispc);
+
+// `WeightVariables` is a generated struct, so we cannot realistically add derivable traits to it.
+// Because of this we disable the clippy warning.
+#[allow(clippy::derivable_impls)]
+impl Default for downsample_ispc::WeightVariables {
+    fn default() -> Self {
+        Self {
+            src_center: Default::default(),
+            src_start: Default::default(),
+            src_end: Default::default(),
+        }
+    }
+}
+
+pub(crate) struct WeightCollection {
+    ispc_representation: downsample_ispc::WeightCollection,
+
+    // Keep these because we need to keep them in memory
+    _starts: Vec<u32>,
+    _weight_counts: Vec<u32>,
+    _weights: Vec<Rc<Vec<f32>>>,
+    _weights_ptrs: Vec<*const f32>,
+}
+
+impl WeightCollection {
+    pub(crate) fn new(mut weights: Vec<CachedWeight>) -> Rc<Self> {
+        let (starts, counts): (Vec<_>, Vec<_>) = weights
+            .iter()
+            .map(|w| (w.start, w.coefficients.len() as u32))
+            .unzip();
+
+        let weights = weights
+            .drain(..)
```

Comment:
Would `into_iter()` (should move the `Vec` and its contents) work here, and allow you to drop the `mut`?